python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,1010 @@
+# coding: utf-8
+import os
+import sys
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.tests.conftest import *  # noqa
+from ray.dag import InputNode, MultiOutputNode, ClassMethodNode
+from ray.dag.dag_node_operation import (
+    DAGNodeOperationType,
+    DAGOperationGraphNode,
+    DAGNodeOperation,
+)
+from ray.dag.compiled_dag_node import _select_next_nodes, CompiledDAG, CompiledTask
+import torch
+from typing import List, Dict, Tuple
+from dataclasses import dataclass, field
+from collections import deque, defaultdict
+from ray.actor import ActorHandle
+
+if sys.platform != "linux" and sys.platform != "darwin":
+    pytest.skip("Skipping, requires Linux or Mac.", allow_module_level=True)
+
+USE_GPU = bool(os.environ.get("RAY_PYTEST_USE_GPU", 0))
+
+
+@dataclass
+class PipelineConfig:
+    """
+    pp_size: Number of pipeline parallel workers.
+    num_micro_batches: Number of micro-batches.
+    """
+
+    pp_size: int
+    num_micro_batches: int
+
+
+@dataclass
+class PipelineUnit:
+    """
+    op: Operation type (FWD or BWD).
+    pp_rank: Pipeline parallel rank.
+    batch_id: Batch ID.
+    uid: Unique ID for the pipeline unit.
+    """
+
+    op: str
+    pp_rank: int
+    batch_id: int
+    uid: str = field(init=False, repr=False)
+
+    def __post_init__(self):
+        self.uid = f"{self.op}_rank-{self.pp_rank}_batch-{self.batch_id}"
+
+    def __repr__(self) -> str:
+        return self.uid
+
+
+def generate_1f1b_schedule(config: PipelineConfig) -> List[List[PipelineUnit]]:
+    """
+    Args:
+        config: Pipeline configuration.
+    Returns:
+        schedule: List of pipeline units for 1F1B pipeline parallelism. Each
+            inner list represents the schedule for a pipeline parallel worker.
+    """
+    pp_size = config.pp_size
+    num_micro_batches = config.num_micro_batches
+
+    schedule = []
+    for pp_rank in range(config.pp_size):
+        warm_up_batches = pp_size - pp_rank
+        main_1f1b_batches = num_micro_batches - warm_up_batches
+        cool_down_batches = num_micro_batches - main_1f1b_batches
+
+        rank_schedule = []
+        bwd_batch_id = fwd_batch_id = 0
+
+        for _ in range(warm_up_batches):
+            rank_schedule.append(PipelineUnit("FWD", pp_rank, fwd_batch_id))
+            fwd_batch_id += 1
+
+        for _ in range(main_1f1b_batches):
+            rank_schedule.append(PipelineUnit("BWD", pp_rank, bwd_batch_id))
+            bwd_batch_id += 1
+            rank_schedule.append(PipelineUnit("FWD", pp_rank, fwd_batch_id))
+            fwd_batch_id += 1
+
+        for _ in range(cool_down_batches):
+            rank_schedule.append(PipelineUnit("BWD", pp_rank, bwd_batch_id))
+            bwd_batch_id += 1
+        schedule.append(rank_schedule)
+    return schedule
+
+
+class PipelineModel:
```

Comment:
I added some comments for PipelineModel, PipelineUnit, and PipelineConfig. I will discuss this with @woshiyyya tomorrow because it was written by Yunxuan.